lib/build.ts

```diff
@@ -187,3 +194,173 @@ export async function runBuildTask(
+//          -> Result: Docker builds, with a warning if the images do not match the
+//                     current platform.
+//
+//      + Some V1 manifests:
+//          -> Result: Docker builds, assuming that the platform matches where the build
+//                     is occurring.
+//
+//      NOTE!  Even when a build succeeds, if Docker is making assumptions then it is possible that
+//      the image will cause an execution error on the target platform if the architecture is not
+//      correct.
+//
+//   So what do we do?
+//
+//      + In the case of all images having a V2 manifest, we can simply pass the platform flag.
+//
+//      + In the case of having some V1 manifests ...
+//          - If we have balenalib images, we know that they have been updated.  Throw an error
+//            and tell the user to update
+//          - If we do not have balenalib images, don't pass the platform flag.  Let Docker
+//            close its eyes and hope for the best.  This opens the possibility for an exec
+//            error, but also allows users to continue to use v1 images if they want to / need to.
+//
+async function checkAllowDockerPlatformHandling(
+	task: BuildTask,
+	docker: Dockerode,
+): Promise<boolean> {
+	let imageReferences: string[];
+
+	const checkHasV1Manifest = async (repository: string) => {
+		try {
+			const manifest = await getManifest(docker.modem, repository);
+			return manifest.Descriptor.mediaType === MEDIATYPE_MANIFEST_V1;
+		} catch {
+			// eat the exception... yummy!
+		}
+
+		console.log(
+			`Image manifest data unavailable for ${repository}. Platform support will be assumed.`,
+		);
+		// at this time, treat any errors as false (no v1 manifest)
+		// This will cause the `--platform` flag to get passed as if the image has
+		// platform support.
+		return false;
+	};
+
+	if (task.imageName) {
+		imageReferences = [task.imageName + !!task.tag ? `:${task.tag}` : ''];
+	} else {
+		if (!task.dockerfile) {
+			// Sanity check
+			console.log('Build task does not have an associated Dockerfile');
+			return false;
+		}
+		const { DockerfileParser } = await import('dockerfile-ast');
+		const parsedDockerfile = DockerfileParser.parse(task.dockerfile);
+		const dockerInstructions = parsedDockerfile.getInstructions();
+		const fromInstructions = dockerInstructions.filter(
+			(inst) => inst.getKeyword() === 'FROM',
+		);
+		if (fromInstructions.length === 0) {
+			// Sanity check
+			console.log(
+				'Build task associated Dockerfile does no reference any images',
+			);
+			return false;
+		}
+		imageReferences = [];
+		fromInstructions.forEach((inst) => {
+			const fromLineArguments = inst.getArguments();
+			if (fromLineArguments.length === 0) {
+				// bail out, no image specified?
+				return;
+			}
+
+			let imgRef = fromLineArguments[0].getValue();
+
+			// hacky way to ignore the --platform flag in the FROM line
+			// In that case, just get the next argument :-p
+			if (imgRef.startsWith('--platform')) {
```

Comment:
At first I thought this would break if the `--platform=linux/arm/v7` was replaced with `--platform linux/arm/v7` but the equals seems to be required according to the docker daemon.
```
FROM --platform linux/arm/v7 alpine:3.12.8
```
```
Error response from daemon: dockerfile parse error line 1: FROM requires either one or three arguments
```
What you have here should work albeit a little hacky as you said.